keps/sig-network/20190603-EndpointSlice-API.md

```diff
@@ -0,0 +1,355 @@
+| Total Bytes Transmitted  | O(PN)                 | O(BN)                           | O(N)                         |
+|                          | ~1KB * 10 = 10KB      | ~1KB * 10 = 10KB                | ~1KB * 10 = 10KB             |
+
+
+### Rolling Update
+
+|                          | Endpoints                   | 100 Endpoints per EndpointSlice | 1 Endpoint per EndpointSlice |
+|--------------------------|-----------------------------|---------------------------------|------------------------------|
+| # of writes              | O(P)                        | O(P)                            | O(P)                         |
+|                          | 20                          | 20                              | 20                           |
+| Size of API object       | O(P)                        | O(B)                            | O(1)                         |
+|                          | ~1KB                        | ~1KB                            | ~1KB                         |
+| # of watchers per object | O(N)                        | O(N)                            | O(N)                         |
+|                          | 10                          | 10                              | 10                           |
+| # of total watch event   | O(NP)                       | O(NP)                           | O(NP)                        |
+|                          | 10 * 20                     | 10 * 20                         | 10 * 20                      |
+| Total Bytes Transmitted  | O(P^2N)                     | O(NPB)                          | O(NP)                        |
+|                          | ~1KB * 10 * 20 = 200KB      | ~1KB * 10 * 20 = 200KB          | ~1KB * 10 * 20 = 200KB       |
+
+
+## Implementation
+
+### Requirements
+
+- Persistence (Minimal Churn of Endpoints)
+
+Upon service endpoint changes, the # of object writes and disruption to ongoing connections should be minimal. 
+
+- Handling Restarts & Failures
+
+The producer/consumer of EndpointSlice must be able to handle restarts and recreate state from scratch with minimal change to existing state.
+ 
+
+### EndpointSlice Controller
+
+A new EndpointSlice Controller will be added to `kube-controller-manager`. It will manage the lifecycle EndpointSlice instances derived from services.  
+```
+Watch: Service, Pod ==> Manage: EndpointSlice
+```
+
+#### Workflows
+On Service Create/Update/Delete:
+- `syncService(svc)`
+
+On Pod Create/Update/Delete: 
+- Reverse lookup relevant services
+- For each relevant service, 
+  - `syncService(svc)`
+
+`syncService(svc)`:
+- Look up selected backend pods
+- For each pod 
+  - If pod is already added into EndpointSlice and Status is correct
+    - Skip
+  - If Pod needs to be added:
+    - If  all EndpointSlice objects has reached the MaxEndpointThreshold. Create a new EndpointSlice object.
+    - Find the EndpointSlice with the lowest number of endpoint, and add it into it. 
```

Comment:
"lowest number of endpoints" gives you a redistribution problem below. You want to completely fill up some and let others empty out as much as possible, which will make them cheaper to redistribute. So, I think you want to add to the one with the highest number of endpoints less than the maximum.

Reply:
+1

Reply:
I see. Okay. I will adjust the algorithm here.